code/scripting/api/libs/ui.cpp

```diff
@@ -2479,5 +2481,309 @@ ADE_FUNC(closePause, l_UserInterface_PauseScreen, nullptr, "Makes sure everythin
+	"Callsign - the name of the message sender, "
+	"Message - the message text, "
+	"Mode - the mode where 0 is normal, 1 is private message, 2 is channel switch, 3 is server message, 4 is MOTD",
+	"table",
+	"A table of chat strings")
+{
+	SCP_UNUSED(L);
+
+	using namespace luacpp;
+
+	LuaTable chats = LuaTable::create(L);
+
+	for (size_t i = 0; i < Multi_pxo_chat.size(); i++) {
+		SCP_list<chat_line>::iterator line = Multi_pxo_chat.begin();
+		std::advance(line, i);
+		
+		char callsign[CALLSIGN_LEN];
+		char text[MAX_CHAT_LINE_LEN];
+
+		char* spacePos = std::strchr(line->text, ' ');
+
+		if (spacePos != nullptr) {
+			// Calculate the length of the callsign
+			int len = spacePos - line->text;
+
+			// Copy the callsign
+			std::strncpy(callsign, line->text, len);
+			callsign[len] = '\0';
+
+			// Copy the text
+			std::strcpy(text, spacePos + 1); // Copy from the character after spacePos
+		}
+
+		int mode;
+		switch (line->mode) {
+		case CHAT_MODE_PRIVATE:
+			mode = 1;
+			break;
+		case CHAT_MODE_CHANNEL_SWITCH:
+			mode = 2;
+			break;
+		case CHAT_MODE_SERVER:
+			mode = 3;
+			break;
+		case CHAT_MODE_MOTD:
+			mode = 4;
+			break;
+		default:
+			mode = 0;
+			break;
+		}
+
+		auto item = luacpp::LuaTable::create(L);
+
+		item.addValue("Callsign", luacpp::LuaValue::createValue(Script_system.GetLuaSession(), callsign));
+		item.addValue("Message", luacpp::LuaValue::createValue(Script_system.GetLuaSession(), text));
+		item.addValue("Mode", luacpp::LuaValue::createValue(Script_system.GetLuaSession(), mode));
+
+		chats.addValue(i + 1, item);
+	}
+
+	return ade_set_args(L, "t", chats);
+}
+
+ADE_FUNC(sendChat, l_UserInterface_MultiPXO, "string", "Sends a string to the current channel's IRC chat", nullptr, nullptr)
+{
+	const char* msg;
+	if (!ade_get_args(L, "s", &msg))
+		return ADE_RETURN_NIL;
+
+	multi_pxo_chat_send(msg);
+
+	return ADE_RETURN_NIL;
+}
+
+ADE_FUNC(getPlayers, l_UserInterface_MultiPXO, nullptr, "Gets the entire player list as a table of strings", "string[]", "A table of player names")
+{
+	SCP_UNUSED(L);
+
+	using namespace luacpp;
+
+	LuaTable chats = LuaTable::create(L);
+
+	for (size_t i = 0; i < Multi_pxo_players.size(); i++) {
+		chats.addValue(i + 1, Multi_pxo_players[i]);
+	}
+
+	return ade_set_args(L, "t", chats);
+}
+
+ADE_FUNC(getPlayerChannel, l_UserInterface_MultiPXO, nullptr, "Searches for a player and returns if they were found and the channel they are on. Channel is an empty string if channel is private or player is not found.", "string string", "The response string and the player's channel")
+{
+	const char* plr_name;
+	if (!ade_get_args(L, "s", &plr_name))
+		return ADE_RETURN_NIL;
+
+	SCP_string cmd;
+	sprintf(cmd, "/WHOIS %s", plr_name);
+
+	multi_pxo_chat_send(cmd.c_str());
+
+	// We have to process the pxo api in order for the response to come back
+	while (stricmp(User_req_channel, "") == 0) {
+		if (Multi_pxo_connected) {
+			multi_pxo_api_process();
+		}
+	}
```

Comment:
Check `Multi_pxo_connected` first thing in the function since nothing here works otherwise.

And unless this is all running in a separate thread the `while` loop here is going to block everything until it fills in `User_req_channel`. If there is an error or the user isn't on a channel or the user is/goes offline then that will never happen. But ideally you should just use `GetChannelByUser()` calls here instead of inventing something new.

Also you need to make a call to `os_sleep(10)` between calls to `multi_pxo_api_process()` to give the system time to handle network stuff.

Reply:
I have a vague memory of trying `GetChannelByUser()` but ran into some issue relating to how the UI through SCPUI runs things differently. But I dunno. It's been so long since I wrote all this that I really don't remember anything concrete anymore and I don't currently have a ton of motivation to dive back in and figure it all out again. I've made the easy change you requested. I guess the rest will have to wait until I feel up to getting neck deep in the multi code again.

Reply:
Alright, I dove back into this and remembered I set it up the way I did originally so that the Lua side was forced to wait for a valid response. I still think that's the better way here given how this is probably going to be used in the UI but I changed it to use GetChannelByUser() and the Lua script will have to be smart enough to handle looping or threading as necessary.

Reply:
I meant to use GetChannelByUser(), but otherwise still do it the same way. That way you use the existing code for the request and error handling. So something like...
```
char *channel = nullptr;

GetChannelByUser(plr_name);

while (channel == nullptr) {
    os_sleep(10);
    multi_pxo_api_process();
    channel = GetChannelByUser(nullptr);
}
```
Then have your previous checks below that but using `channel` instead of `User_req_channel`.

Basically the first call to GetChannelByUser() initializes it and sends the request (return value doesn't matter on first call). Then you call it repeatedly with nullptr and check the return value until it's non-null. The return value will then either be "-1" if an error occurred, or the channel name. 

With your recent change it actually wouldn't work at all. Sorry for the confusion there.